### PR TITLE
APP-2902 Authentication API reference documentation

### PR DESCRIPTION
**Pull request by symphony-thibault**
### Ticket
[APP-2902](https://perzoinc.atlassian.net/browse/APP-2902)

### Description
Reference documentation around the Authentication configuration and API. 

### Dependencies
List the other pull requests that should be merged before/along this one.

### Checklist
- [x] Referenced a ticket in the PR title and in the corresponding section
- [x] Filled properly the description and dependencies, if any
- [x] Unit tests updated or added
- [x] Javadoc added or updated
- [x] Updated the documentation in [docs folder](../docs)

### CODE REVIEW

File: docs/authentication.md [hunk 1]
```diff
@@ -1,52 +1,145 @@
...
+### App Authentication using Client Certificate
 
-In order to authenticate your extension app, you need to have a proper `app` field defined in your configuration,
-see [authentication](authentication.md) for more information. We support both RSA and certificate authentication.
+Required `config.yaml` setup: 
+```yaml
+host: acme.symphony.com
+app:
+    appId: app-id
+    certificatePath: /path/to/certificate.cer
```
Comment (symphony-hong) [inline docs/authentication.md:102]: The certificate for extension app should be in pem format.
https://developers.symphony.com/extension/docs/application-authentication#section-prerequisites-for-client-certificate-authentication

File: symphony-bdk-core/src/main/java/com/symphony/bdk/core/SymphonyBdk.java [hunk 2]
```diff
@@ -142,12 +149,31 @@ public AuthSession obo(String username) throws AuthUnauthorizedException {
...
 
   /**
-   * Extension App Authenticate by app token
-   * @param appToken App Token
-   * @return Extension App authentication session
+   * Returns the {@link ExtensionAppAuthenticator}.
+   *
+   * @return the {@link ExtensionAppAuthenticator}
+   */
+  public ExtensionAppAuthenticator appAuthenticator() {
```
Comment (symphony-hong) [inline symphony-bdk-core/src/main/java/com/symphony/bdk/core/SymphonyBdk.java:156]: why you want to expose the AppAuthenticator from SymphonyBdk?

**Review by symphony-elias — APPROVED** [review]
LGTM

File: docs/authentication.md (hunk 1 above)
Comment (symphony-elias) [reply docs/authentication.md:102]: IMHO documentation is not so clear. We use PEM format when registering the public certificate in the admin console. However, we use a p12 file which is in DER format when authenticating.

Comment (symphony-hong) [reply docs/authentication.md:102]: @symphony-elias no, I used the crt format when registering the public certificate in the admin console and the p12 when authenticating.

File: symphony-bdk-core/src/main/java/com/symphony/bdk/core/SymphonyBdk.java (hunk 2 above)
Comment (symphony-thibault) [reply symphony-bdk-core/src/main/java/com/symphony/bdk/core/SymphonyBdk.java:156]: That's not perfect I agree, but before that the `SymphonyBdk` was not providing a way to implement the Circle Of Trust. We might have to revamp a bit this feature

**Review by symphony-hong — APPROVED** [review]
LGTM 👍

**Review by symphony-youri — COMMENTED** [review]
In the examples we use loadFromSymphonyDir is it something that is meant for us to use for development or also for BDK users (if so did we introduce it somewhere in the doc)?

There is a typo that could be fixed before merging, my other comments are potential improvements or open questions

File: symphony-bdk-examples/bdk-spring-boot-example/src/test/java/com/symphony/bdk/examples/spring/arch/ArchitectureTest.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+ * <p>
+ *    Description of the tests:
+ *    <ol>
+ *      <li>methods annotated with {@link Slash} must have 1 parameter of type {@link CommandContext}</li>
+ *      <li>{@link AbstractActivity} classes must be annotated with {@link Service} or {@link Component}</li>
+ *    </ol>
+ * </p>
+ */
+@AnalyzeClasses(packages = "com.symphony.bdk.examples.spring")
+public class ArchitectureTest {
```
[the remark this answers is not in the document]
Comment (symphony-thibault) [reply symphony-bdk-examples/bdk-spring-boot-example/src/test/java/com/symphony/bdk/examples/spring/arch/ArchitectureTest.java:28]: ~`800ms` to run this test class

File: symphony-bdk-examples/bdk-spring-boot-example/src/test/java/com/symphony/bdk/examples/spring/arch/ArchitectureTest.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+  @ArchTest
+  void all_slash_methods_must_have_1_command_context_parameter(JavaClasses classes) {
```
[the remark this answers is not in the document]
Comment (symphony-thibault) [reply symphony-bdk-examples/bdk-spring-boot-example/src/test/java/com/symphony/bdk/examples/spring/arch/ArchitectureTest.java:31]: Not yet, this was just an example. We can however introduce it in the `core` module as well. Checking `@API` annotation usage is a really good idea 👍 (cf. `APP-3096`)

File: docs/authentication.md [hunk 5]
```diff
@@ -1,52 +1,145 @@
...
+Required `config.yaml` setup: 
+```yaml
+host: acme.symphony.com
+bot:
+    username: bot-username
+    privateKeyPath: /path/to/rsa/private-key.pem
+```
 
-To be updated...
+### Bot authentication using Client Certificate
```
[the remark this answers is not in the document]
Comment (symphony-thibault) [reply docs/authentication.md:25]: That's a very vague topic at Symphony as some of our biggest customers still want to rely on certificate authentication. Until a clear decision is taken, I'd prefer to not mention that this is deprecated. As well, in our public API doc we don't mention that this auth method is deprecated either

File: docs/authentication.md [hunk 6]
```diff
@@ -1,52 +1,145 @@
...
+```java
+public class Example {
+
+  public static void main(String[] args) throws BdkConfigException, AuthInitializationException, AuthUnauthorizedException {
+
+    // setup SymphonyBdk facade object
+    final SymphonyBdk bdk = new SymphonyBdk(loadFromSymphonyDir("config.yaml"));
+    
+    final AuthSession oboSessionUsername = bdk.obo("user.name");
+    final AuthSession oboSessionUserId = bdk.obo(123456789L);
```
[the remark this answers is not in the document]
Comment (symphony-thibault) [reply docs/authentication.md:140]: Correct, I added an example using the `StreamService`. I actually noticed that our API is not really `OBO-ready`. For instance we cannot send a message using OBO. 

@symphony-elias or @symphony-hong: could you create stories to OBO available in other services?

File: docs/authentication.md [hunk 7]
```diff
@@ -1,52 +1,145 @@
...
+bot:
+    username: bot-username
+    certificatePath: /path/to/certificate.cer
+    certificatePassword: YourCertificatePassword
+```
 
-On Behalf Of (OBO) is a pattern that enables developers to perform operations on behalf of a Symphony end-user. [`Getting started with OBO](https://developers.symphony.com/restapi/docs/get-started-with-obo)
+### Configuration priority
+If you configure both private key and certificate within the same `config.yaml`, note that the certificate will have
+a higher priority over the RSA configuration.
```
[the remark this answers is not in the document]
Comment (symphony-thibault) [reply docs/authentication.md:39]: No real use case for that, I just document the existing behaviour. 

@symphony-hong or @symphony-elias could you create a story to implement @symphony-youri's suggestion?

**Review by symphony-youri — APPROVED** [review]

File: docs/authentication.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (symphony-hong) [reply docs/authentication.md:39]: Ticket created: https://perzoinc.atlassian.net/browse/APP-3098

**Pull request merged**